livekit-rtc/livekit/rtc/rpc.py

```diff
@@ -0,0 +1,109 @@
+
+from typing import Optional, Dict, Union, ClassVar
+from enum import IntEnum
+from ._proto import rpc_pb2 as proto_rpc
+from dataclasses import dataclass
+
+
+@dataclass
+class RpcInvocationData:
+    """Data passed to method handler for incoming RPC invocations
+
+    Attributes:
+        request_id (str): The unique request ID. Will match at both sides of the call, useful for debugging or logging.
+        caller_identity (str): The unique participant identity of the caller.
+        payload (str): The payload of the request. User-definable format, typically JSON.
+        response_timeout (float): The maximum time the caller will wait for a response.
+    """
+
+    request_id: str
+    caller_identity: str
+    payload: str
+    response_timeout: float
+
+
+class RpcError(Exception):
+    """
+    Specialized error handling for RPC methods.
+
+    Instances of this type, when thrown in a method handler, will have their `message`
+    serialized and sent across the wire. The caller will receive an equivalent error on the other side.
+
+    Build-in types are included but developers may use any string, with a max length of 256 bytes.
+    """
+
+    class ErrorCode(IntEnum):
+        APPLICATION_ERROR = 1500
+        CONNECTION_TIMEOUT = 1501
+        RESPONSE_TIMEOUT = 1502
+        RECIPIENT_DISCONNECTED = 1503
+        RESPONSE_PAYLOAD_TOO_LARGE = 1504
+        SEND_FAILED = 1505
+
+        UNSUPPORTED_METHOD = 1400
+        RECIPIENT_NOT_FOUND = 1401
+        REQUEST_PAYLOAD_TOO_LARGE = 1402
+        UNSUPPORTED_SERVER = 1403
+        UNSUPPORTED_VERSION = 1404
+
+    ErrorMessage: ClassVar[Dict[ErrorCode, str]] = {
+        ErrorCode.APPLICATION_ERROR: "Application error in method handler",
+        ErrorCode.CONNECTION_TIMEOUT: "Connection timeout",
+        ErrorCode.RESPONSE_TIMEOUT: "Response timeout",
+        ErrorCode.RECIPIENT_DISCONNECTED: "Recipient disconnected",
+        ErrorCode.RESPONSE_PAYLOAD_TOO_LARGE: "Response payload too large",
+        ErrorCode.SEND_FAILED: "Failed to send",
+        ErrorCode.UNSUPPORTED_METHOD: "Method not supported at destination",
+        ErrorCode.RECIPIENT_NOT_FOUND: "Recipient not found",
+        ErrorCode.REQUEST_PAYLOAD_TOO_LARGE: "Request payload too large",
+        ErrorCode.UNSUPPORTED_SERVER: "RPC not supported by server",
+        ErrorCode.UNSUPPORTED_VERSION: "Unsupported RPC version",
+    }
+
+    def __init__(
+        self,
+        code: Union[int, "RpcError.ErrorCode"],
+        message: str,
+        data: Optional[str] = None,
+    ):
+        """
+        Creates an error object with the given code and message, plus an optional data payload.
+
+        If thrown in an RPC method handler, the error will be sent back to the caller.
+
+        Error codes 1001-1999 are reserved for built-in errors (see RpcError.ErrorCode for their meanings).
+        """
+        super().__init__(message)
+        self.code = code
+        self.message = message
+        self.data = data
```

Comment:
We should prefer to expose those using @property. Otherwise they're not referenced on the docs and they're mutable

Reply:
Good call, thank you.